src/core/expr/fexpr_count_countna.cc

```diff
@@ -0,0 +1,160 @@
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+//------------------------------------------------------------------------------
+#include "column/const.h"
+#include "column/latent.h"
+#include "column/countna.h"
+#include "column/countna_no_args.h"
+#include "documentation.h"
+#include "expr/fexpr_reduce_unary.h"
+#include "expr/eval_context.h"
+#include "expr/workframe.h"
+#include "python/xargs.h"
+#include "stype.h"
+namespace dt {
+namespace expr {
+
+template<bool COUNTNA>
+class FExpr_CountNA : public FExpr_ReduceUnary {
+  public:
+    using FExpr_ReduceUnary::FExpr_ReduceUnary;
+
+
+    std::string name() const override {
+      return COUNTNA? "countna"
+                    : "count";
+    }
+
+    Column evaluate1(Column&& col, const Groupby& gby, bool is_grouped) const {
+      SType stype = col.stype();
+      switch (stype) {
+        case SType::VOID:
+        case SType::BOOL:
+        case SType::INT8:
+          return make<int8_t>(std::move(col), SType::INT64, gby, is_grouped);
+        case SType::INT16:
+          return make<int16_t>(std::move(col), SType::INT64, gby, is_grouped);
+        case SType::DATE32: 
+        case SType::INT32:
+          return make<int32_t>(std::move(col), SType::INT64, gby, is_grouped);
+        case SType::TIME64: 
+        case SType::INT64:
+          return make<int64_t>(std::move(col), SType::INT64, gby, is_grouped);
+        case SType::FLOAT32:
+          return make<float>(std::move(col), SType::INT64, gby, is_grouped);
+        case SType::FLOAT64:
+          return make<double>(std::move(col), SType::INT64, gby, is_grouped);
+        case SType::STR32:
+        case SType::STR64:
+          return make<CString>(std::move(col), SType::INT64, gby, is_grouped);                
+        default:
+          throw TypeError()
+            << "Invalid column of type `" << stype << "` in " << repr();
+      }
+    }
+
+    template <typename T>
+    Column make(Column&& col, SType stype, const Groupby& gby, bool is_grouped) const {
+      if (is_grouped) {
+        return Column(new Latent_ColumnImpl(new Count_ColumnImpl<T, COUNTNA, true>(
+          std::move(col), stype, gby
+        )));
+      } else {
+        return Column(new Latent_ColumnImpl(new Count_ColumnImpl<T, COUNTNA, false>(
+          std::move(col), stype, gby
+        )));
```

Comment:
Don't need `stype` here either, because `Count_ColumnImpl` knows it is always `SType::INT64`.

Reply:
The original `reduce_unary` derives its stype from the col: 

```py
template <typename T>
class ReduceUnary_ColumnImpl : public Virtual_ColumnImpl {
  protected:
    Column col_;
    Groupby gby_;

  public:
    ReduceUnary_ColumnImpl(Column &&col, const Groupby& gby)
      : Virtual_ColumnImpl(gby.size(), col.stype()),
        col_(std::move(col)),
        gby_(gby)
    {
      xassert(col_.can_be_read_as<T>());
    }


    ColumnImpl *clone() const override {
      return new ReduceUnary_ColumnImpl(Column(col_), Groupby(gby_));
    }


    size_t n_children() const noexcept override {
      return 1;
    }


    const Column &child(size_t i) const override {
      xassert(i == 0);
      (void)i;
      return col_;
    }
};
```

I tweaked it such that you can explicitly pass an SType instead - this is handy for the count function

```py
template <typename T_IN, typename T_OUT>
class ReduceUnary_ColumnImpl : public Virtual_ColumnImpl {
  protected:
    Column col_;
    Groupby gby_;
    

  public:
    ReduceUnary_ColumnImpl(Column &&col, SType stype, const Groupby& gby)
      : Virtual_ColumnImpl(gby.size(), stype),
        col_(std::move(col)),
        gby_(gby)
    {
      xassert(col_.can_be_read_as<T_IN>());
    }


    ColumnImpl *clone() const override {
      return new ReduceUnary_ColumnImpl(Column(col_), this->stype(), Groupby(gby_));
    }


    size_t n_children() const noexcept override {
      return 1;
    }


    const Column &child(size_t i) const override {
      xassert(i == 0);
      (void)i;
      return col_;
    }
};
```

Are you suggesting I revert to your initial setup, and just use a different class altogether for the `count` function? maybe adapt the `FExpr_Count_Rows` function instead?